File: Editorial/3101-3200/3161_Block_Placement_Queries/CSharp/2_Fenwick_Tree.cs
public class Solution {
    private int[] bt;

    private void Update(int x, int v) {
        for (; x < bt.Length; x += x & -x) {
            bt[x] = Math.Max(bt[x], v);
        }
    }

    private int Query(int x) {
        int res = 0;
        for (; x > 0; x -= x & -x) {
            res = Math.Max(res, bt[x]);
        }
        return res;
    }

    public IList<bool> GetResults(int[][] queries) {
        int mx = 50000;

        SortedSet<int> st = new SortedSet<int> { 0, mx };
        foreach (int[] q in queries) {
            if (q[0] == 1) {
                st.Add(q[1]);
            }
        }

        List<int> sortedList = new List<int>(st);
        bt = new int[mx + 1];
        int pre = 0;
        for (int i = 0; i < sortedList.Count; i++) {
            int x = sortedList[i];
            if (x == 0) {
                continue;
            }
            Update(x, x - pre);
            pre = x;
        }

        List<bool> ans = new List<bool>();
        for (int i = queries.Length - 1; i >= 0; i--) {
            int[] q = queries[i];
            if (q[0] == 2) {
                int x = q[1];
                int sz = q[2];

                int idx = sortedList.BinarySearch(x);
                if (idx < 0) {
                    idx = ~idx - 1;
                }

                int preVal = sortedList[idx];
                int maxSpace = Query(preVal);
                maxSpace = Math.Max(maxSpace, x - preVal);
                ans.Add(maxSpace >= sz);
            } else {
                int x = q[1];

                int idx = sortedList.BinarySearch(x);
                int preVal = idx > 0 ? sortedList[idx - 1] : 0;
                int nxt = idx < sortedList.Count - 1 ? sortedList[idx + 1] : mx;

                Update(nxt, nxt - preVal);
                sortedList.RemoveAt(idx);
            }
        }

        ans.Reverse();
        return ans;
    }
}

File: Editorial/3101-3200/3161_Block_Placement_Queries/TypeScript/2_Fenwick_Tree.ts
function getResults(queries: number[][]): boolean[] {
    const mx = 50000;
    const st: number[] = [0, mx];

    const bisectLeft = (arr: number[], x: number): number => {
        let lo = 0,
            hi = arr.length;
        while (lo < hi) {
            const mid = (lo + hi) >>> 1;
            if (arr[mid] < x) lo = mid + 1;
            else hi = mid;
        }
        return lo;
    };

    const insort = (arr: number[], x: number): void => {
        const idx = bisectLeft(arr, x);
        arr.splice(idx, 0, x);
    };

    for (const q of queries) {
        if (q[0] === 1) {
            insort(st, q[1]);
        }
    }

    const bt: number[] = new Array(mx + 1).fill(0);
    const update = (x: number, v: number): void => {
        while (x <= mx) {
            if (v > bt[x]) bt[x] = v;
            x += x & -x;
        }
    };

    const query = (x: number): number => {
        let res = 0;
        while (x > 0) {
            if (bt[x] > res) res = bt[x];
            x -= x & -x;
        }
        return res;
    };

    let pre = 0;
    for (const x of st) {
        if (x === 0) {
            continue;
        }
        update(x, x - pre);
        pre = x;
    }

    const ans: boolean[] = [];
    for (let i = queries.length - 1; i >= 0; i--) {
        const q = queries[i];
        if (q[0] === 2) {
            const x = q[1],
                sz = q[2];
            const idx = bisectLeft(st, x);
            let preVal: number;

            if (idx < st.length && st[idx] === x) {
                preVal = x;
            } else {
                preVal = st[idx - 1];
            }

            let maxSpace = query(preVal);
            maxSpace = Math.max(maxSpace, x - preVal);
            ans.push(maxSpace >= sz);
        } else {
            const x = q[1];
            const idx = bisectLeft(st, x);
            const preVal = st[idx - 1];
            const nxt = st[idx + 1];

            update(nxt, nxt - preVal);
            st.splice(idx, 1);
        }
    }

    return ans.reverse();
}
